gms/endpoint_state.cc

```diff
@@ -65,15 +65,12 @@ locator::host_id endpoint_state::get_host_id() const noexcept {
 }
 
 std::optional<locator::endpoint_dc_rack> endpoint_state::get_dc_rack() const {
-    if (auto app_state = get_application_state_ptr(application_state::DC)) {
-        std::optional<locator::endpoint_dc_rack> ret;
-        ret->dc = app_state->value();
-        if ((app_state = get_application_state_ptr(application_state::RACK))) {
-            ret->rack = app_state->value();
-            if (ret->dc.empty() || ret->rack.empty()) {
-                on_internal_error_noexcept(logger, format("Node {} has empty dc={} or rack={}", get_host_id(), ret->dc, ret->rack));
-            }
-            return ret;
+    if (const auto* dc_state = get_application_state_ptr(application_state::DC)) {
+        const auto* rack_state = get_application_state_ptr(application_state::RACK);
+        if (dc_state->value().empty() || !rack_state || rack_state->value().empty()) {
+            on_internal_error_noexcept(logger, format("Node {} has empty dc={} or rack={}", get_host_id(), dc_state->value(), rack_state ? rack_state->value() : "(null)"));
```

Comment:
Why crash here?

Reply:
It's only in testing.  Production will just print an error and return `nullopt`.
The reason is that the endpoint state should never contain partial dc/rack state and/or empty values for dc/rack.

Reply:
Also, it's preexisting,.  This pr only adds a term to the condition by checking also the presences of the RACK state, not only if it's empty.

Reply:
How does production know not to crash?

Reply:
Is "on_internal_error_noexcept" the new abort(), that I dedicated a github issue to its eradication?
Why not use "on_internal_error" which *does* throw, and let only the code that called this function stop instead of the entire database?

Reply:
The code was copied, not new.

Reply:
> How does production know not to crash?

It aborts only with the --abort-on-intrrnal-error flag, which is set in testing and clear by default in production.



> Is "on_internal_error_noexcept" the new abort(), that I dedicated a github issue to its eradication? Why not use "on_internal_error" which _does_ throw, and let only the code that called this function stop instead of the entire database?

Not exactly...  on_fatal_internal_error is the new `abort()`, just with a proper message and oneline backtrace.
`on_internal_error_except` prints the error and backtrace, and then maybe abort, depending on the flag mentioned above, otherwise it returns normally, without throwing (hence the _noexcept suffix).
The reason it is used here is that there is a sane recovery from this condition which is to return nullopt.

> The code was copied, not new.

right. we can consider changing that, but it could be done in a follow up PR.